File: 프로그래머스/unrated/159993. 미로 탈출/미로 탈출.py
from collections import deque


def solution(maps):
    def bfs(s, e):
        x, y = points[s]
        q = deque([(x, y, 0)])
        visited = [[False] * len(maps[0]) for _ in range(len(maps))]
        visited[x][y] = True
        while q:
            x, y, d = q.popleft()
            if maps[x][y] == e:
                return d
            for i in range(4):
                nx, ny = x + dx[i], y + dy[i]
                if nx < 0 or nx >= len(maps) or ny < 0 or ny >= len(maps[0]) or visited[nx][ny] or maps[nx][ny] == 'X':
                    continue
                visited[nx][ny] = True
                q.append((nx, ny, d + 1))
        return 0
    
    dx = (-1, 0, 1, 0)
    dy = (0, 1, 0, -1)
    points = {}
    
    for i in range(len(maps)):
        for j in range(len(maps[0])):
            if maps[i][j] in ('S', 'E', 'L'):
                points[maps[i][j]] = (i, j)
                
    s_to_l, l_to_e = bfs('S', 'L'), bfs('L', 'E')
    return s_to_l + l_to_e if s_to_l and l_to_e else -1

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {

    private static final int[] dx = {-1, 0, 1, 0};
    private static final int[] dy = {0, 1, 0, -1};

    public int solution(String[] maps) {
        int n = maps.length;
        int m = maps[0].length();

        int[] start = null;
        int[] lever = null;
        int[] end = null;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                char c = maps[i].charAt(j);
                if (c == 'S') {
                    start = new int[]{i, j};
                } else if (c == 'L') {
                    lever = new int[]{i, j};
                } else if (c == 'E') {
                    end = new int[]{i, j};
                }
            }
        }

        int sToL = bfs(start, 'L', maps, n, m);
        if (sToL == -1) {
            return -1;
        }

        int lToE = bfs(lever, 'E', maps, n, m);
        if (lToE == -1) {
            return -1;
        }

        return sToL + lToE;
    }

    private int bfs(int[] start, char target, String[] maps, int n, int m) {
        Queue<int[]> q = new ArrayDeque<>();
        boolean[][] visited = new boolean[n][m];

        q.offer(new int[]{start[0], start[1], 0});
        visited[start[0]][start[1]] = true;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0], y = cur[1], dist = cur[2];

            if (maps[x].charAt(y) == target) {
                return dist;
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx < 0 || ny < 0 || nx >= n || ny >= m) {
                    continue;
                }
                if (visited[nx][ny] || maps[nx].charAt(ny) == 'X') {
                    continue;
                }

                visited[nx][ny] = true;
                q.offer(new int[]{nx, ny, dist + 1});
            }
        }

        return -1;
    }
}